### PR TITLE
fix cmake build for mingw and msvc.

### PR DESCRIPTION
**Pull request by aliaspider**
also includes some small fixes.
the videos in the mingw build are currently broken, it might be because of the different ffmpeg libs.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -50,6 +50,10 @@ include(ccache)
 # Remove soon?
 set(USE_FFMPEG ON)
 
+if(MINGW AND NOT DEFINED USE_SYSTEM_FFMPEG)
```
Comment (hrydgard) [inline CMakeLists.txt]: SYSTEM_FFMPEG is generally not a good idea, we should probably make a custom mingw build if people really want to use it. I don't think people do use it a lot though, so a comment would suffice.

File: Windows/ppsspp.rc [hunk 2]
```diff
@@ -228,8 +266,26 @@ BEGIN
     CONTROL         "",IDC_GEDBG_MATRICES,"SysListView32",LVS_ALIGNLEFT | LVS_SHOWSELALWAYS | LVS_REPORT | WS_BORDER | WS_TABSTOP,7,7,217,86
 END
 
-
+#ifdef __MINGW32__
```
Comment (hrydgard) [inline Windows/ppsspp.rc]: Why does this need to be different on mingw?

File: Windows/ppsspp.rc [hunk 3]
```diff
@@ -110,9 +110,47 @@ END
 //
 // Version
 //
-
+#ifdef __MINGW32__
```
Comment (hrydgard) [inline Windows/ppsspp.rc]: How come mingw needs a special path here?

File: Windows/mingw_defines.h [hunk 4]
```diff
@@ -0,0 +1,234 @@
+#pragma once
```
Comment (hrydgard) [inline Windows/mingw_defines.h]: Is this file needed, why?

**Review by hrydgard — CHANGES_REQUESTED** [review]
Neat. I'm not super familiar with mingw so I've sprinkled a few questions around.

File: Windows/ppsspp.rc (hunk 2 above)
Comment (aliaspider) [reply Windows/ppsspp.rc]: because the resource compiler doesn't support utf-16.
this is actually not even a good solution since it doesn't even treat the file as utf-8.
I can't figure out a better solution for this so this was the only workaround I could come up with.

File: Windows/mingw_defines.h (hunk 4 above)
Comment (aliaspider) [reply Windows/mingw_defines.h]: this is needed because the headers from the dx9sdk folder aren't fully compatible with mingw.

File: CMakeLists.txt (hunk 1 above)
Comment (aliaspider) [reply CMakeLists.txt]: adding the mingw static libs would be better yeah.
the bundeled msvc libs generate missing symbol errors when used with mingw, but I can change that to a comment if that's preferred

File: Windows/ppsspp.rc (hunk 3 above)
Comment (aliaspider) [reply Windows/ppsspp.rc]: see next comment.

File: CMakeLists.txt (hunk 1 above)
Comment (hrydgard) [reply CMakeLists.txt]: Nah, just add a comment here saying that this is why you are using system ffmpeg in this case.

File: Windows/ppsspp.rc (hunk 2 above)
Comment (hrydgard) [reply Windows/ppsspp.rc]: Oh, darn. But please add a comment saying that, just on the line above.

File: Windows/mingw_defines.h (hunk 4 above)
Comment (hrydgard) [reply Windows/mingw_defines.h]: Again, add a comment to the code saying so, for the benefit of future readers :)

**Comment by hrydgard** [comment]
Getting build errors in Travis, on clang linux:

```/home/sergio/Projects/PPSSPP/ppsspp-git/ffmpeg/libavformat/id3v2.c:952: undefined reference to `uncompress'```

File: CMakeLists.txt [hunk 5]
```diff
@@ -419,7 +430,13 @@ target_link_libraries(Common Ext::Snappy)
 
 if(WIN32)
 	include_directories(dx9sdk/Include)
-	target_link_libraries(Common winmm d3d9 dxguid dsound)
+   include_directories(dx9sdk/Include/DX11)
+   if(MINGW)
+      set(CMAKE_RC_COMPILE_OBJECT "<CMAKE_RC_COMPILER> -O coff <INCLUDES> <SOURCE> <OBJECT>")
+      # required when using the dx9sdk include paths
+      add_definitions(-include ${CMAKE_CURRENT_SOURCE_DIR}/Windows/mingw_defines.h)
+      add_definitions(-D_WIN32_WINNT=0x0601)
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Hmm, can we get this from stdafx.h?  Also we seem to use 0x600 elsewhere - could we use the same here?

-[Unknown]

File: GPU/Vulkan/TextureCacheVulkan.cpp [hunk 6]
```diff
@@ -182,7 +182,11 @@ void TextureCacheVulkan::DeviceLost() {
 	}
 
 	samplerCache_.DeviceLost();
-	vulkan_->Delete().QueueDeleteSampler(samplerNearest_);
+
+	if(samplerNearest_) {
+		vulkan_->Delete().QueueDeleteSampler(samplerNearest_);
+		samplerNearest_ = VK_NULL_HANDLE;
```
[the remark this answers is not in the document]
Comment (aliaspider) [reply GPU/Vulkan/TextureCacheVulkan.cpp]: ok I will remove that.

File: CMakeLists.txt (hunk 5 above)
Comment (aliaspider) [reply CMakeLists.txt]: 0x601 is the minimum required version to compile some files (touch API usage mostly).

Comment (unknownbrackets) [reply CMakeLists.txt]: Ah - that's a good catch, we should change that, then.

-[Unknown]

**Comment by aliaspider** [comment]
fixed build error and updated code based on current reviews.

**Review by hrydgard — APPROVED** [review]

**Comment by hrydgard** [comment]
The appveyor build still errors out for some reason...

**Comment by aliaspider** [comment]
ok going to check it again

**Review by unknownbrackets — APPROVED** [review]
